run ci_complete

s-hertel opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -211,19 +212,22 @@ def find_module(self, fullname, path=None):
             # TODO: log attempt to load context
             return None
 
+        return importlib.util.spec_from_loader(fullname, loader)
```
It's likely not necessary for our use, but you could pass `origin` here also.

```
origin=loader.get_filename(loader._fullname)
```

One thing to note is that `_AnsibleInternalRedirectLoader` doesn't implement `get_filename` so could just be added to return `None`.

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -531,6 +531,13 @@ def _validate_final(self):
+        AnsibleCollectionConfig.on_collection_load.fire(collection_name=collection_name, collection_path=os.path.dirname(module.__file__))
+
+    def create_module(self, spec):
+        return self.load_module(spec.name)
```
This probably needs to change slightly, to perhaps not rely on `load_module`, however, maybe some of the code from `load_module` could move here, and then `load_module` could depend on `create_module`.

To explain a little bit, `pep489` is for a multi stage import system.  Where the stages are effectively:

1. `create_spec` (and loader)
2. `create_module` - this just creates the module container, but doesn't exec any code
3. `exec_module` - this execs the code

As it stands, `load_module` has historically done 2 and 3, and in the current implementation still calls `exec`

If you look at `importlib.module_from_spec` and `_new_module`, you will see the empty container is created by doing `type(sys)(spec.name)` effectively.

So `_new_or_existing_module` does too much for the context of `create_module`.

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -531,6 +531,13 @@ def _validate_final(self):
 
+    def exec_module(self, module):
+        collection_name = '.'.join(self._split_name[1:3])
+        AnsibleCollectionConfig.on_collection_load.fire(collection_name=collection_name, collection_path=os.path.dirname(module.__file__))
```
I think, for this, we basically just need the `exec` stuff from `load_module`.  But I haven't tested all of the pieces put together at the same time, so I could be wrong.

s-hertel commented:
/azp run

mkrizek commented:
!waffling support:community

s-hertel commented:
bot_skip

webknjaz commented:
/azp run

webknjaz commented:
/azp run

s-hertel commented:
@webknjaz you can ignore this draft btw, not sure how to prevent it showing up in your queue. I have the bot skipped on it so ci_verified is never removed.I use it to run tests if I'm on another branch locally, and I'll never merge anything via this PR.

webknjaz commented:
@s-hertel this one appears in the queue not because of the labels but because of the pending CI. Sometimes integrations send updates to GH but it fails to update its internal state and "thinks" that CI checks haven't been reported in full (even though there have and show up as green). For these, we still restart the CI so it'll disappear from the queue.

@mattclay maybe we could exclude this from search by PR number?

s-hertel commented:
/azp run

mattclay commented:
> @mattclay maybe we could exclude this from search by PR number?

@webknjaz Sorry, I missed this somehow. How about a dedicated label for PRs to ignore when checking CI? That way we don't need to update the queries when the list changes. I thought about ignoring drafts, but I think that would be too broad of an exclusion, since about 33% of our open PRs are drafts.

s-hertel committed:
Ensure the task result's _ansible_delegated_vars['ansible_host'] value is templated, so the callback plugin's host_label uses the rendered value.

s-hertel committed:
Or, add a mechanism for the connection plugin to return the host, port, and user

Add host_label option to local connection

Remove malformed ssh connection variables

Improve connection debug messages by using the connection plugin values instead of play context (which may not be templated)

document historical-ish option as 2.20 option, although the user facing change is a bugfix not feature

s-hertel committed:
Add changelog fragment for fixed display issue

Add changelog fragment for variables literally named "delegated_vars['ansible_ssh_host']" and "delegated_vars['ansible_host']"

s-hertel committed:
Update unit test

s-hertel committed:
ci_complete

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -772,13 +772,23 @@ def _execute_internal(self, templar: TemplateEngine, variables: dict[str, t.Any]
+
+            # TODO: consider deprecating or adding a better mechanism to pass
+            # variables to the results side. The comment above isn't really
+            # true since cvars is not templated.
```
I think it is fine to pass them w/o templating, so they can examine the sources 'as they were', what we should also do is pass the 'resolved' options with the connection itself

also we should look to deprecate callbacks that rely on variables directly, but we should ensure that the interface for getting the resolved information from the connection object is working. 

in any case, the current fix seems to be fine, specially for backporting, but we want to get defined objects/interfaces for callbacks going forward.

s-hertel replied:
I'm not sure if there's a need. I overlooked that these are filtered out here https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_result.py#L202, and after looking through the ansible-collections org, I don't see any callback plugins using anything other than "ansible_host" in the delegated variables.

How does https://github.com/ansible/ansible/pull/85397 look for backporting? I removed the PlayContext changes.

bcoca commented on lib/ansible/plugins/connection/local.py:
```diff
@@ -12,6 +12,14 @@
+              - The host name potentially displayed by callback plugins or debug messages.
+            vars:
+              - name: inventory_hostname
+              - name: ansible_host
```
should we make this something shared? it seems something people would want to customize for most plugins

s-hertel replied:
Maybe a `remote_addr` option could be convenient? I thought the option name `host_label` would be better for local specifically though. This new option was solely to fix the local connection output with `-vvv` which can display a template string (since pc remote_addr may be one).

bcoca commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -741,6 +737,24 @@ def _persistence_controls(b_command: list[bytes]) -> tuple[bool, bool]:
+    def user(self) -> str:
+        if not (remote_user := self.get_option("remote_user")):
+            raise NotImplementedError
+        return remote_user
```
something to move to base class? probalby with play_context fallback?

s-hertel replied:
I don't think so, since the base methods do not use pc port or user, only the remote_addr. The ssh plugin sets these properties internally, and I was just trying to figure out if the ssh plugin doesn't have set_options reliably set (but I believe it does).

s-hertel committed:
ci_complete

s-hertel committed:
put a pin in pc train of thought

s-hertel committed:
ci_complete

s-hertel committed:
ci_complete

s-hertel committed:
ci_complete
